### ISSUE TITLE
sql: refactor casts to be more like postgresql

### ISSUE THREAD

**Issue by otan** [cockroachdb/cockroach#55094]
Our current method of handling casts and operators are a bit of a mess.

Casts are defined in a massive switch statement, and their volatility in a separate map (validCasts).

This causes a lot of confusion, especially with `AdjustValueFor[Column]Type`, whose responsibilities are placed on both INSERT and UPDATE time, instead of simply parse and cast time. In an ideal state, this "stays" as parse/cast features, and items we store should always fit the column width as the type is kept the same.

We should refactor casts and parse to be more like Postgres: there should be a map that roughly goes `map[fromType oid.Oid]map[toType oid.Oid]castOps` that defines casts from and to. They should be able to re-use the parse function -- strings should automatically be populated into these cast maps using the parse definitions.

**Comment by otan** [comment]
>  It seems like OID -> OID could cause a lot of duplicate code and boilerplate.

Well, it doesn't if you have implicit/assignment casts :). OID -> OID is most `postgres`-like, and the Family:Family stuff is a huge mess when you look at the various different times of `String`-family OIds we have. However, unless we do the type system refactor I think Family:Family makes sense (for now).

> They should be able to re-use the parse function -- strings should automatically be populated into these cast maps using the parse definitions.

I phrased this badly, let me rephrase. This is in the context of string casts, which IIUC goes through ParseAndRequireString anyway.

In PG, type I/O input/output functions depend on `typin` and `typout` in `pg_type`: https://github.com/postgres/postgres/blob/537ca68dbb2463f7b1c44e9466b8fbdd7505b2e1/src/backend/utils/adt/int.c#L265-L283

What I believe I meant was that the casts should re-use the `ParseD....` functions, but I think ::string and string:: related casts already have some variation of this magic.

**Comment by mgartner** [comment]
> Well, it doesn't if you have implicit/assignment casts :)

I don't follow. OID -> OID would require a map that contains casting details for each OID pair. Family -> Family would by definition be a smaller map.

> However, unless we do the type system refactor I think Family:Family makes sense (for now).

Sounds good, I'll give it a shot.

**Comment by otan** [comment]
> Family -> Family would by definition be a smaller map.

smaller map, a lot more if conditions ;)
